From 9d328bb5d0e0e92772a431ea113bfe248dd863d6 Mon Sep 17 00:00:00 2001
From: Jani Lusikka <jani.lusikka@gmail.com>
Date: Tue, 14 Feb 2017 19:49:44 +0800
Subject: [PATCH] [PATCH] Fix ownership for files/folders created by
 ExternalStorageProvider on ext4/f2fs

By forcing ExternalStorageProvider to use the sdcard FUSE layer for access,
ext4/f2fs sdcards will be treated with the correct media_rw:media_rw
permissions instead of the current app permissions.

Change-Id: I2cbb97727ebe5af88a879490732925406a165af5
---
 .../src/com/android/externalstorage/ExternalStorageProvider.java  | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/packages/ExternalStorageProvider/src/com/android/externalstorage/ExternalStorageProvider.java b/packages/ExternalStorageProvider/src/com/android/externalstorage/ExternalStorageProvider.java
index 10e5dcc..df96c11 100644
--- a/packages/ExternalStorageProvider/src/com/android/externalstorage/ExternalStorageProvider.java
+++ b/packages/ExternalStorageProvider/src/com/android/externalstorage/ExternalStorageProvider.java
@@ -214,6 +214,14 @@ public class ExternalStorageProvider extends DocumentsProvider {
             }
             root.path = volume.getInternalPathForUser(userId);
             try {
+
+                // Force all ext4/f2fs sdcard acccess through the sdcard FUSE layer to ensure the
+                // correct permissions for accessing files and directories.
+                if (volume.getType() == VolumeInfo.TYPE_PUBLIC && root.visiblePath != null
+                    && ("ext4".equals(volume.fsType) || "f2fs".equals(volume.fsType))) {
+                    root.path = root.visiblePath;
+                }
+
                 root.docId = getDocIdForFile(root.path);
             } catch (FileNotFoundException e) {
                 throw new IllegalStateException(e);
-- 
2.1.4

